core/internal/testutils/pgtest/txdb.go

```diff
@@ -80,19 +80,18 @@ func (d *txDriver) Open(dsn string) (driver.Conn, error) {
 		d.db = db
 	}
 	c, exists := d.conns[dsn]
-	if !exists {
+	if !exists || !c.Open() {
 		tx, err := d.db.Begin()
 		if err != nil {
 			return nil, err
 		}
-		c = &conn{tx: tx}
+		c = &conn{tx: tx, opened: 1}
 		d.conns[dsn] = c
 		c.removeSelf = func() error {
 			return d.deleteConn(c)
 		}
 		d.conns[dsn] = c
 	}
-	c.opened++
```

Comment:
Instead of racing with the decrement and check in Close(), call Open() which tries to increment while holding the lock.

Reply:
:+1: